src/systems/particle_emitter2/ParticleEmitter2.cc

```diff
@@ -0,0 +1,289 @@
+      "emitter system. The requested command will be ignored.\n";
+  }
+}
+
+//////////////////////////////////////////////////
+bool ParticleEmitter2Private::EmittersService(
+    ignition::msgs::ParticleEmitter_V &_res)
+{
+  _res.Clear();
+
+  // Lock and wait for an iteration to be run and fill the state
+  std::unique_lock<std::mutex> lock(this->serviceMutex);
+
+  this->serviceRequest = true;
+  bool success = this->serviceCv.wait_for(lock, 5s, [&]
+  {
+    return !this->serviceRequest;
+  });
+
+  if (success)
+    _res.CopyFrom(this->serviceMsg);
+  else
+    ignerr << "Timed out waiting for state" << std::endl;
+
+  return success;
+}
+
+//////////////////////////////////////////////////
+ParticleEmitter2::ParticleEmitter2()
+  : System(), dataPtr(std::make_unique<ParticleEmitter2Private>())
+{
+}
+
+//////////////////////////////////////////////////
+void ParticleEmitter2::Configure(const Entity &_entity,
+    const std::shared_ptr<const sdf::Element> & /*_sdf*/,
+    EntityComponentManager &_ecm,
+    EventManager & /*_eventMgr*/)
+{
+  // World
+  const components::Name *name = _ecm.Component<components::Name>(_entity);
+  if (name == nullptr)
+  {
+    ignerr << "World with id: " << _entity
+           << " has no name. ParticleEmitter2 cannot create transport topics\n";
+    return;
+  }
+
+  std::string emittersService = "/world/" + name->Data() + "/particle_emitters";
+  if (this->dataPtr->node.Advertise(emittersService,
+        &ParticleEmitter2Private::EmittersService, this->dataPtr.get()))
+  {
+    ignmsg << "Serving particle emitter information on ["
+      << emittersService << "]" << std::endl;
+  }
+  else
+  {
+    ignerr << "Something went wrong, failed to advertise [" << emittersService
+           << "]" << std::endl;
+  }
+}
+
+//////////////////////////////////////////////////
+void ParticleEmitter2::PreUpdate(const ignition::gazebo::UpdateInfo &_info,
+    ignition::gazebo::EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("ParticleEmitter2::PreUpdate");
+
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+
+  // Create particle emitters
+  _ecm.EachNew<components::ParticleEmitter>(
+    [&](const Entity &_entity,
+        const components::ParticleEmitter *_emitter)->bool
+      {
+        std::string topic;
+
+        // Get the topic information from the header, which is currently a
+        // hack to avoid breaking the particle_emitter.proto message.
+        if (_emitter->Data().has_header())
+        {
+          for (const auto data : _emitter->Data().header().data())
+          {
+            if (data.key() == "topic" && !data.value().empty())
+            {
+              topic = data.value(0);
+            }
+          }
+        }
+
+        // If a topic has not been specified, then generate topic based
+        // on the model, link, and emitter names.
+        if (topic.empty())
+        {
+          std::string emitterScopedName =
```

Comment:
I think you can generate the topic string by doing:

```
        topic = "/" + removeParentScope(removeParentScope(
               scopedName(_entity, _ecm, "/", true), "/"), "/") + "/cmd";
```

this should also handle multiple levels of nested models.

Reply:
Thanks. I added a helper function that encapsulates this logic in 1b623241952a80fa7d7bffd9961949ee6b21577c.